### PR TITLE
Do not resolve wildcards if upper bound is too non-specific

### PR DESCRIPTION
**Pull request by yawkat**
This is a draft fix for #4118. I will explain two approaches I tried.

[The first approach](https://github.com/FasterXML/jackson-databind/commit/9257b230cd4d1bbb0fc6d0a9423e281601e2728e) was as follows. When parsing a `ParameterizedType`, if I see a wildcard, I also construct the type for the upper bound of the type variable. Then, if the wildcard upper bound is a supertype of the type variable upper bound, I use the type variable upper bound instead. (The `!javaType.hasGenericTypes()` check is there to make the supertype checking safe.)

The problem with this is that it leads to a stack overflow. Before this attempt, it would produce the type `Tree<Object>`. Now it tries to build the type `Tree<Tree<Tree<Tree<…>>>>` which breaks.

The only way out of this conundrum that I could see was raw types. Basically the workaround from #4118. Now there was a choice to make. Either `Tree<?>` could resolve to `Tree<Tree>`, or to `Tree`.

Resolving to `Tree<Tree>` has a problem with correctness. If the type variable is declared to have an upper bound of `X<Y>`, it is *not* necessarily safe to discard the `Y` and just replace the wildcard with `X`. I didn't use this approach for that reason.

The second raw type approach, resolving to `Tree`, is also problematic. If the input is actually `Tree<X, ?>`, we cannot discard `X` in order to make the wildcard work. At the same time, There's no way to express a type `Tree<X, >` where the second type variable is unbound.

Finally, I decided to go with the last approach. I added a method to `TypeBindings` to "unbind" a specific variable. If a wildcard is detected, and its upper bound is non-specific, I unbind it again. This leads to downstream deserialization logic using the bound of the type variable. Fortunately this existing logic can deal with the recursiveness of the type.

This approach also required an optimization of the `rawClass` method in order to avoid recursive type resolution.

### CODE REVIEW

**Comment by pjfanning** [comment]
Could this be targeted at 2.16 branch instead?

**Comment by yawkat** [comment]
yea you're right.

also a disclaimer: i think this change will affect the type produced by *most* wildcards, so it has high potential for adverse impact. i assume that in most cases with wildcards, the lower bound matches that of the type parameter (both Object), in which case the type will now be dropped but *should* trigger the same deser code path later on. so in principle the behavior should be the same, but edge cases are likely.

**Comment by cowtowncoder** [comment]
I am not sure if this might be going down the wrong rathole: is the problem really the recursive type definition?
Resolving self-referential type by truncation (as done currently) has seemed work fine so far, and not be the root cause for this particular problem.

To me it seems rather that when resolving `TestAttribute<?>` of Field (etc) declaration, the fact that `TestAttribute` class specifies type parameter bounds is not taken into account.
So, we resolve `?` directly as `java.lang.Object`, instead of considering bounds it has.

I may be wrong with my analysis of course, but I think I can come up with a test case that at least demonstrates specific failing case from outside, without trying to directly test `TypeBindings`

**Comment by yawkat** [comment]
The problem is not the recursive type definition, it's just a *major*
complicating factor. Without it, the fix would be easy. I agree with your
analysis, and it's essentially what this PR fixes—it just has to do it in
an awkward way (by unbinding) because of the possibility of a recursive
type definition.

The test case in this PR, while not identical to the original issue, should
accurately reproduce the problem at the deserialization level.

File: src/main/java/com/fasterxml/jackson/databind/type/TypeFactory.java [hunk 1]
```diff
@@ -17,6 +8,34 @@
 import com.fasterxml.jackson.databind.util.LRUMap;
 import com.fasterxml.jackson.databind.util.LookupCache;
 
+import java.lang.reflect.Array;
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/type/TypeFactory.java:11]: Let's leave 5+ imports wild-carded. Fully expanding brings lots of fluff...

**Comment by cowtowncoder** [comment]
Ok. I can follow the explanation and agree that there's the risk of regression; but that also this should resolve #4118. So will merge in 2.16 and hope for the best. :)

**Review by cowtowncoder — APPROVED** [review]

**Comment by pjfanning** [comment]
Can we consider again if this is a good idea for a 2.x release? It is the most likely cause of broken tests in jackson-module-scala. Could we at least consider introducing a Feature that allows users to opt in to this behaviour. If we think semver is a good idea, then this change should be non-default and users can choose to enable it.

**Comment by pjfanning** [comment]
I've tried to hack jackson-module-scala to work with the changed v2.16 code. This is not going well. 
If I use `objectMapper.readValue(json, new TypeReference<Either<?, String>>(){})` and then use objectMapper.readValue(json, new TypeReference<Either<String, ?>>(){}) - jackson-databind 2.16 mixes the 2 up and uses the first typereference. They are not the same, note that the type parameter order is different.

jackson-databind now only attaches one type binding and ignores the `?`. This seems fundamentally wrong to me. We need to have the 2 type bindings, even if we need to create a special type binding for the `?`.

**Comment by yawkat** [comment]
@pjfanning that sounds fixable. the intention of this patch is to have only one of the two params bound. this *should* work fine because the binding is associated with the type var name, so it shouldnt be misused for the other variable. if it is, that sounds like a bug we could resolve without going back on this entirely.

i can look at this tomorrow

**Comment by pjfanning** [comment]
> @pjfanning that sounds fixable. the intention of this patch is to have only one of the two params bound. this _should_ work fine because the binding is associated with the type var name, so it shouldnt be misused for the other variable. if it is, that sounds like a bug we could resolve without going back on this entirely.
> 
> i can look at this tomorrow

Thanks @yawkat - if you have trouble reproducing the issue, I can try to create a Java specific scenario. The scenario I described above is a Scala scenario but I have used Java syntax in the write up. The problem with creating a Java specific scenario is that it will probably need a lot of code (creating a new Java class similar to scala.util.Either and then writing a deserializer for it). There is a Vavr class and a Jackson module for Vavr that might possibly have this issue.

**Comment by pjfanning** [comment]
I found that vavr-jackson is also broken. Probably similar issue to jackson-module-scala. I forked vavr-jackson and created a branch on my fork that fixes a compile issue and when I run the tests the Either tests are broken. https://github.com/pjfanning/vavr-jackson/tree/jackson-2.16

**Comment by pjfanning** [comment]
@yawkat I have some broken tests in https://github.com/FasterXML/jackson-databind/pull/4149 -- targeted at 2.15 branch where they work but I've tried them in 2.16 and they fail. Getting them to pass in 2.16 may not fix my jackson-module-scala problems but it there is a good chance that a fix that fixes these tests will also help my Scala test cases.

**Comment by cowtowncoder** [comment]
Ok I am sorry this became what looks like a mess. If there are things that got broken I really hope they can be resolved -- but if not, I'd rather revert the change than add any feature flags. This isn't (in my opinion) something that users have information to decide, given that breakage relies on rather low-level details.

I do think resolution would be necessary before `2.16.0-rc1` so I will hold off for release until we know how to proceed here.

**Pull request merged**